src/arrow/schema.rs

```diff
@@ -0,0 +1,992 @@
+      other => Err(ArrowError(
+        format!("Unable to convert parquet logical type {}", other)))
+    }
+  }
+  
+  // Functions for group types.
+  
+  fn to_struct(&self) -> Result<Option<DataType>> {
+    match self.schema.as_ref() {
+      Type::PrimitiveType {..} => panic!("This should not happen."),
+      Type::GroupType {
+        basic_info: _,
+        fields
+      } => {
+        fields.iter()
+          .map(|field_ptr| self.clone_with_schema(field_ptr.clone()).to_field())
+          .collect::<Result<Vec<Option<Field>>>>()
+          .map(|result| result.into_iter()
+            .filter_map(|f| f)
+            .collect::<Vec<Field>>())
+          .map(|fields| {
+            if fields.is_empty() {
+              None
+            } else {
+              Some(DataType::Struct(fields))
+            }
+          })
+      }
+    }
+  }
+  
+  fn to_list(&self) -> Result<Option<DataType>> {
+    match self.schema.as_ref() {
+      Type::PrimitiveType {..} => panic!("This should not happen."),
+      Type::GroupType {
+        basic_info: _,
+        fields
+      } if fields.len() == 1 => {
+        let list_item = fields.first().unwrap();
+        let item_converter = self.clone_with_schema(list_item.clone());
+        
+        let item_type = match list_item.as_ref() {
+          Type::PrimitiveType {
+            ..
+          }  => {
+            if item_converter.is_repeated() {
+              item_converter.to_primitive_type_inner().map(|dt| Some(dt))
+            } else {
+              Err(ArrowError(
+                "Primitive element type of list must be repeated.".to_string()))
+            }
+          },
+          Type::GroupType {
+            basic_info: _,
+            fields
+          }  => {
+            if fields.len() > 1 {
+              item_converter.to_struct()
+            } else if fields.len() == 1 && list_item.name() != "array"
+              && list_item.name() != format!("{}_tuple", self.schema.name()) {
+
+              let nested_item = fields.first().unwrap();
+              let nested_item_converter = self.clone_with_schema(nested_item.clone());
+
+              nested_item_converter.to_data_type()
+            } else {
+              item_converter.to_struct()
+            }
+          }
+        };
+        
+        item_type.map(|opt| opt.map(|dt| DataType::List(Box::new(dt))))
+      },
+      _ => Err(ArrowError(
+        "Group element type of list can only contain one field.".to_string()))
+    }
+  }
+}
+
+
+
+#[cfg(test)]
+mod tests {
+  use std::rc::Rc;
+  
+  use schema::types::GroupTypeBuilder;
+  use schema::types::SchemaDescriptor;
+  use basic::Type as PhysicalType;
+  use basic::LogicalType;
+  use basic::Repetition;
+  use schema::types::PrimitiveTypeBuilder;
+  
+  use arrow_lib::datatypes::Field;
+  use arrow_lib::datatypes::DataType;
+
+  use super::parquet_to_arrow_schema;
+  use super::parquet_to_arrow_schema_by_columns;
+  
+  macro_rules! make_parquet_type {
```

Comment:
Could you also check test utils? We might already have a function there to generate schemas.

Reply:
Thanks. I found a parse which can convert texts to type. I'll change tests to use that.